[Model] Add support for Jina Embeddings   V4

realsigridjin opened the pull request:
## Purpose

  This PR adds support for the Jina Embeddings V4 model
  (`jinaai/jina-embeddings-v4-vllm-retrieval`) in vLLM, enabling multimodal embeddings for text and image inputs.

FIX #20463 

  ## Test Plan
* Run the comprehensive test suite for Jina V4
     ```bash
     pytest
  tests/models/multimodal/test_jina_embeddings_v4.py -v
```

## Documentation Update

  The following documentation has been added:

  1. Model Registry (vllm/model_executor/models/registry.py): Added JinaVLForEmbedding entry
  2. Supported Models (docs/source/models/supported_models.rst): Added Jina Embeddings V4 to the multimodal embeddings section
  3. Example Script (examples/offline_inference/jina_embeddings_v4.py):
  4. Implementation Guide  (JINA_EMBEDDINGS_V4_IMPLEMENTATION.md): Detailed
  technical documentation

realsigridjin committed:
feat: jina support

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,356 @@
+                output = seq_states.mean(dim=0)
+            
+            # Normalize
+            output = F.normalize(output, p=2, dim=-1)
```
mistake. fixed here. https://github.com/vllm-project/vllm/pull/20802/commits/d7d6b6048f4223c18a318488df04e7ce482361f4#diff-bc18208ee0cfb6279e0c9ef1df4692b32b16d1d9a8a1eb650867372b9ac4612a

sigridjineth commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,356 @@
+            # Clean up any temporary tensors
+            if 'token_tensor' in locals():
+                del token_tensor
+            torch.cuda.empty_cache()
```
amended here https://github.com/vllm-project/vllm/pull/20802/commits/d7d6b6048f4223c18a318488df04e7ce482361f4#diff-bc18208ee0cfb6279e0c9ef1df4692b32b16d1d9a8a1eb650867372b9ac4612a

DarkLight1337 reviewed (commented):
Thanks for contributing! Can you add this model to the [test registry](https://github.com/vllm-project/vllm/blob/main/tests/models/registry.py) and [supported models page](https://github.com/vllm-project/vllm/blob/main/docs/models/supported_models.md)?

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,363 @@
+            
+            return PoolerOutput(outputs=pooled_outputs)
+            
+        except Exception as e:
```
Why not fail fast?

DarkLight1337 replied:
I prefer always failing fast so we can know about potential issues about using the optimized pooler

sigridjineth replied:
done https://github.com/vllm-project/vllm/pull/20802/commits/5b67b13cd026bb77ba664c2014bbddd1db24b74a

DarkLight1337 commented on vllm/model_executor/models/registry.py:
```diff
@@ -168,6 +168,8 @@
     # input and output. I am adding it here because it piggy-backs on embedding
     # models for the time being.
     "PrithviGeoSpatialMAE": ("prithvi_geospatial_mae", "PrithviGeoSpatialMAE"),
+    # Multimodal embedding model with token-type-aware pooling
+    "JinaVLForEmbedding": ("jina_embeddings_v4", "JinaVLForEmbedding"),
```
Let's move this to multimodal section like Jina-VL

sigridjineth replied:
Will move to multimodal section

DarkLight1337 replied:
The location of the model in the registry doesn't affect which endpoints are enabled. It's based on the model implementation

DarkLight1337 replied:
Oh, just realized that, sorry for the confusion. Let's move the existing Jina-VL back to embedding section then

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,363 @@
+
+
+# Triton kernel for optimized vision token extraction
+if HAS_TRITON:
```
How much is the performance increase using triton? cc @Isotr0py @imkero

Isotr0py replied:
If this triton kernel is only used in pooler, I think the performance improvement will be very little. But it would be best to have a performance benchmarks first.

Isotr0py commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,363 @@
+
+    def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]):
+        """Load weights with validation and error handling."""
+        try:
+            loader = AutoWeightsLoader(self)
+            loaded_weights = loader.load_weights(weights, mapper=self.weight_mapper)
+            logger.info(f"Successfully loaded {len(loaded_weights)} weight tensors")
+            return loaded_weights
+        except Exception as e:
+            logger.error(f"Error loading weights: {e}")
+            raise
```
```suggestion
        loader = AutoWeightsLoader(self)
        loaded_weights = loader.load_weights(weights, mapper=self.weight_mapper)
        logger.info(f"Successfully loaded {len(loaded_weights)} weight tensors")
        return loaded_weights
```
Please clean up the `try: ... except: ...` statement used in debug.

realsigridjin committed:
refactor: fail fast

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

realsigridjin committed:
refactor: exceptions

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,345 @@
+
+    def _extract_token_ids_safe(
+        self, 
+        pooling_metadata: PoolingMetadata
```
We should handle the case of `V1PoolingMetadata` as well. Refer to the existing logic in `pooler.py`

sigridjineth replied:
amended https://github.com/vllm-project/vllm/pull/20802/commits/d644c5c4997dca3f4d9d87bf6cebed3fd97e4827

realsigridjin committed:
refactor: improve jina embeddings v4 model

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,345 @@
+            return token_ids_list, seq_ids
+            
+        except Exception as e:
+            logger.error(f"Error extracting token IDs: {e}")
```
amended https://github.com/vllm-project/vllm/pull/20802/commits/d644c5c4997dca3f4d9d87bf6cebed3fd97e4827

realsigridjin committed:
refactor: oom

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,345 @@
+                hidden_states, token_ids_list, prompt_lens
+            )
+        except RuntimeError as e:
+            if "out of memory" in str(e).lower():
```
amended https://github.com/vllm-project/vllm/pull/20802/commits/48c4ea489694854b4cf93625145a11b53902b604

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,363 @@
+VISION_END_TOKEN_ID = 151653
+
+# Maximum sequence length for safety
+MAX_SEQUENCE_LENGTH = 512 * 1024  # 512K tokens
```
Could you elaborate on this setting? vLLM already validates the prompt length by comparing it against `max_model_len`, so I'm not sure whether we really need this

sigridjineth replied:
the MAX_SEQUENCE_LENGTH check is indeed  redundant given that vLLM already  performs validation against  max_model_len. I've removed this manual  check to rely on the framework's  built-in validation, which simplifies  the code and avoids duplication of  logic.

thanks for pointing this out tho

sigridjineth replied:
the MAX_SEQUENCE_LENGTH check is indeed redundant
thanks for pointing this out tho

realsigridjin committed:
refactor: Validate lengths match

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,359 @@
+        # Validate lengths match
+        if len(token_ids_list) != len(prompt_lens):
+            logger.error(f"Mismatch: {len(token_ids_list)} sequences vs {len(prompt_lens)} lengths")
+            return self._base_pooler(hidden_states, pooling_metadata)
```
amended https://github.com/vllm-project/vllm/pull/20802/commits/714e28365faec8fad1d3ad03086f3746ac8e9976

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,364 @@
+                pooling_metadata, hidden_states.device
+            ).prompt_lens
+        
+        # Validate lengths match
+        if len(token_ids_list) != len(prompt_lens):
+            raise AssertionError(
+                f"Mismatch: {len(token_ids_list)} sequences vs {len(prompt_lens)} lengths"
+            )
```
```suggestion
        # Validate lengths match
        assert len(token_ids_list) == len(prompt_lens), (
            f"Mismatch: {len(token_ids_list)} sequences vs {len(prompt_lens)} lengths" )
```

realsigridjin committed:
refactor: normalize

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,349 @@
+
+            # Normalize and handle potential NaNs by replacing with zeros
+            output = F.normalize(output, p=2, dim=-1)
+            output = torch.nan_to_num(output)
```
Actually this is probably unnecessary as `F.normalize` has `eps` parameter to avoid division by zero

sigridjineth replied:
yup thanks https://github.com/vllm-project/vllm/pull/20802

realsigridjin committed:
refactor: normalize

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

realsigridjin committed:
refactor: review

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,348 @@
+        vision_start_id: tl.constexpr,
+        vision_end_id: tl.constexpr,
+        BLOCK_SIZE: tl.constexpr,
+    ):
```
done https://github.com/vllm-project/vllm/pull/20802/commits/caea1fead89bc4d0c46a5d66a00b2812104986cf

sigridjineth commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,348 @@
+except ImportError:
+    HAS_TRITON = False
+    triton = None
+    tl = None
```
done https://github.com/vllm-project/vllm/pull/20802/commits/caea1fead89bc4d0c46a5d66a00b2812104986cf

realsigridjin committed:
refactor: prehook commits

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

Commit:
fix: Apply isort formatting to jina_embeddings_v4.py

Fixed import statement formatting to comply with isort requirements.
The PoolingMetadata import now has proper line breaks and indentation.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

Commit:
[ci skip-hooks] Formatting attempt(s)

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>
This is a known issue where CI runs formatters on all files, not just changed files.

DarkLight1337 commented:
You can `yapf: disable` the affected lines

Commit:
fix: Resolve yapf/isort conflict with disable comments

As suggested by maintainer, use yapf: disable/enable comments around
the pooling_metadata imports to prevent formatter conflicts. This allows
isort to handle the import formatting while yapf skips these lines.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,308 @@
+
+PoolingMetadata = Union[V0PoolingMetadata, V1PoolingMetadata]
+
+# Triton kernel for optimized vision token extraction
```
Move this comment to the other file

DarkLight1337 commented:
Once https://github.com/vllm-project/vllm/pull/21058 is merged, you also have to update this PR by refactoring the `pooler` call into a class that inherits from `Pooler`

Commit:
refactor: accept review

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented:
@DarkLight1337 okay, will look forward to. when is the expected merge date then?

Isotr0py commented on vllm/model_executor/models/registry.py:
```diff
@@ -170,6 +170,8 @@
     # input and output. I am adding it here because it piggy-backs on embedding
     # models for the time being.
     "PrithviGeoSpatialMAE": ("prithvi_geospatial_mae", "PrithviGeoSpatialMAE"),
+    # Multimodal embedding model with token-type-aware pooling
+    "JinaVLForEmbedding": ("jina_embeddings_v4", "JinaVLForEmbedding"),
```
We should also update `tests/models/registry.py`.

sigridjineth replied:
the `tests/models/registry.py` file was already updated with the jinavlforembedding entry in the initial commit addressing review feedback at lines 319-320, so this review comment is outdated and has been addressed!

Commit:
refactor: address review feedback for Jina embeddings V4

- Switch default pooling backend from Triton to PyTorch due to performance issues
- Remove inappropriate benchmark validation file
- Move example to vision_language_embedding.py
- Add JinaVLForEmbedding to test registry

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

Commit:
refactor: import HAS_TRITON from triton_utils instead of local definition

Import HAS_TRITON directly from vllm.triton_utils to maintain consistency
across the codebase and avoid redundant definitions.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented on vllm/model_executor/layers/pooler.py:
```diff
@@ -15,9 +15,12 @@
     PoolingMetadata as V0PoolingMetadata)
 from vllm.model_executor.pooling_metadata import PoolingTensors
 from vllm.sequence import PoolerOutput, PoolingSequenceGroupOutput
+from vllm.triton_utils import tl, triton
```
did it https://github.com/vllm-project/vllm/pull/20802/commits/0c3f1bd6c7d452e3d316fa69d74b32c61dcf85cc

Commit:
refactor: rename example file to follow existing embedding pattern

Rename vision_language_embedding.py to embed_jina_embeddings_v4.py to
follow the established naming pattern for embedding examples (e.g.,
embed_jina_embeddings_v3.py, embed_matryoshka_fy.py).

Also update the docstring to be more specific about Jina Embeddings V4
multimodal capabilities.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

sigridjineth commented on benchmarks/jina_embeddings_v4_validation.py:
did it https://github.com/vllm-project/vllm/pull/20802/commits/fafd668fe95cdd363e71b764a071d7447fabd00f

Commit:
Merge remote-tracking branch 'origin/main' into jina-support

Commit:
refactor: update JinaVLForEmbedding to comply with new pooling architecture

Update JinaVLForEmbedding to align with PR #21058's pooling model interface:
- Add is_pooling_model = True class attribute
- Create JinaVLPooler class inheriting from Pooler base class
- Move vision-aware pooling logic into JinaVLPooler
- Implement get_pooling_params method returning PoolingParams() for "embed" task
- Replace pooler method with pooler attribute
- Add required imports: PoolingTask, PoolingParams, assert_never

The JinaVLPooler maintains the sophisticated vision-text pooling behavior
while conforming to the new architecture requirements.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,335 @@
+        # Validate inputs
+        if hidden_states is None or hidden_states.numel() == 0:
+            logger.warning("Empty hidden states received")
+            return PoolerOutput(outputs=[])
```
Please use the utility functions from `pooler.py` such as `build_output` and `get_prompt_lens` to avoid duplicate code

sigridjineth replied:
@DarkLight1337 can you confirm whether this has been resolved?

sigridjineth commented:
@DarkLight1337 managed it to merge with main branch

https://github.com/vllm-project/vllm/pull/20802/commits/8e0578a7d2492a8901368adee92ffb1fc6a5030f

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,335 @@
+    def get_pooling_params(self, task: PoolingTask) -> Optional[PoolingParams]:
+        """Return pooling params for embedding task."""
+        if task == "embed":
+            return PoolingParams()
```
Need to set `logits_processing_needs_token_ids`, otherwise the token IDs may not be accessible in V1

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,335 @@
+        # Initialize the vision-aware pooler
+        self.pooler = JinaVLPooler(vllm_config, self.pooling_backend)
+
+        logger.info("Initialized JinaVLForEmbedding with thread-safe pooling")
```
Is this log still relevant?

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,335 @@
+            return PoolingParams()
+
+        # The equalities are split up to keep mypy happy
+        if task == "encode" or task == "classify" or task == "score":
```
`"encode"` should always be supported because it is a more general task that returns the output data without enforcing dimensional constraints.

Commit:
refactor: use pooler utility functions to avoid duplicate code

Use build_output and get_prompt_lens from pooler.py instead of
implementing duplicate logic:
- Replace manual PoolerOutput construction with build_output
- Replace prompt length extraction logic with get_prompt_lens
- Remove unused imports (PoolingSequenceGroupOutput, PoolingTensors)

This addresses the review feedback to avoid duplicate code.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,335 @@
+            return PoolerOutput(outputs=[])
+
+        # Extract token IDs safely from metadata
+        token_ids_list, seq_ids = self._extract_token_ids_safe(
```
`seq_ids` seems unused, can you remove the code that returns it?

Commit:
refactor: address maintainer review comments for JinaVLPooler

Address DarkLight1337's review feedback:
- Set logits_processing_needs_token_ids=True for V1 compatibility in both
  "embed" and "encode" tasks
- Support "encode" task by returning PoolingParams() instead of None
- Update log message from "thread-safe pooling" to "vision-aware pooling"
  to better reflect the actual functionality
- Remove unused seq_ids variable from _extract_token_ids_safe method

These changes ensure proper V1 compatibility and cleaner code structure.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on vllm/model_executor/layers/pooler.py:
```diff
@@ -706,3 +707,41 @@ def forward(
+
+        for i in range(seq_len):
+            token_id = tl.load(token_ids_ptr + seq_start + i)
+            if token_id >= vision_start_id and token_id <= vision_end_id:
```
```suggestion
            if token_id in (vision_start_id, vision_end_id):
```

DarkLight1337 replied:
You applied `hidden_size` in the triton kernel but not in the unoptimized code

DarkLight1337 replied:
See how the official code handles it: https://huggingface.co/jinaai/jina-embeddings-v4/blob/main/modeling_jina_embeddings_v4.py#L228

DarkLight1337 replied:
Can you fix this?

sigridjineth replied:
Hi @DarkLight1337, thanks for highlighting this. reflected again and got that your assessment is correct. there's a positional indexing mistake in the implementation.

sigridjineth replied:
the problem was I think it incorrectly selects only tokens exactly matching the vision_start_id and vision_end_id.

It fails to select intermediate tokens between these markers because it uses direct ID matching instead of positional masking. I have created on the commit to introduce the dedicated `VisionPooler` class that finds the positions of `vision_start_id` and `vision_end_id` via `torch.where` method. the goal is to ensure pooling the entire positional range.

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,326 @@
+                                      device=hidden_states.device)
+
+            # Check for vision tokens
+            vision_mask = ((seq_tokens >= VISION_START_TOKEN_ID) &
+                           (seq_tokens <= VISION_END_TOKEN_ID))
```
Same as L209

Commit:
perf: optimize vision token detection using torch.isin

Implement efficiency improvements suggested by DarkLight1337:
- Consolidate get_pooling_params method for "embed" and "encode" tasks
- Pre-compute vision token IDs tensor in constructor
- Replace range checks with torch.isin for more efficient vision token detection
  at lines 209-210 and 261-262

This reduces redundant code and improves performance when checking for
vision tokens by using optimized tensor operations.

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

Commit:
fix: introducing dedicated VisionPooler class

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,327 @@
+            normalize=True,
+            softmax=False)
+
+    def get_pooling_params(self, task: PoolingTask) -> Optional[PoolingParams]:
+        """Return pooling params for embedding task."""
+        if task == "embed" or task == "encode":
+            return PoolingParams(logits_processing_needs_token_ids=True)
+
+        # The equalities are split up to keep mypy happy
+        if task == "classify" or task == "score":
+            return None
+
+        assert_never(task)
```
Can you merge from main and then apply this update? (Need to update the imports accordingly as well)

```suggestion
    def get_pooling_updates(
        self,
        task: PoolingTask,
    ) -> Optional[PoolingParamsUpdate]:
        # The equalities are split up to keep mypy happy
        if task == "encode" or task == "embed":
            return PoolingParamsUpdate(requires_token_ids=True)

        if task == "classify" or task == "score":
            return None

        assert_never(task)
```

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,327 @@
+
+        return build_output(pooled_tensor)
+
+    def _extract_token_ids_safe(
+            self, pooling_metadata: PoolingMetadata) -> list[array]:
+        """Safely extract token IDs from pooling metadata."""
+        token_ids_list: list[array] = []
+        try:
+            if isinstance(pooling_metadata, V1PoolingMetadata):
+                # For V1, we get token IDs directly
+                for i, num in enumerate(pooling_metadata.prompt_lens):
+                    token_ids = pooling_metadata.prompt_token_ids[
+                        i, :num].tolist()
+                    token_ids_list.append(array('l', token_ids))
+
+                return token_ids_list
+
+            # For V0, we extract from seq_groups and seq_data
+            for seq_group, _ in pooling_metadata.seq_groups:
+                for seq_id in seq_group:
+                    if seq_id not in pooling_metadata.seq_data:
+                        logger.warning("Sequence %s not found in seq_data",
+                                       seq_id)
+                        continue
+
+                    seq_data = pooling_metadata.seq_data[seq_id]
+
+                    # Get prompt token IDs safely
+                    if hasattr(seq_data, 'prompt_token_ids_array'):
+                        token_ids = seq_data.prompt_token_ids_array
+                    elif hasattr(seq_data, '_prompt_token_ids'):
+                        token_ids = seq_data._prompt_token_ids
+                    else:
+                        logger.warning("No token IDs found for sequence %s",
+                                       seq_id)
+                        continue
+
+                    token_ids_list.append(token_ids)
+
+            return token_ids_list
```
The latest main now has `get_prompt_token_ids` in `pooler.py` which can replace this functionality (but note that it outputs `torch.Tensor` instead of `array.array`)

sigridjineth commented on vllm/model_executor/layers/pooler.py:
```diff
@@ -32,6 +32,7 @@ class PoolingType(IntEnum):
     CLS = 2
     STEP = 3
     MEAN = 4
+    VISION = 5
```
I have created this new type of `vision` pooling for PoolingClass.

DarkLight1337 commented on vllm/model_executor/layers/pooler.py:
```diff
@@ -706,3 +790,83 @@ def forward(
         return build_output(scores)
+
+
+class VisionPooler(Pooler):
```
Did you accidentally copy this? The same class appears twice in the file

Commit:
feat: add vision pooling support for jina embeddings v4

Signed-off-by: Sigrid Jin (Sionic AI) <sigrid@sionic.ai>

DarkLight1337 commented on tests/models/pooling/test_jina_embeddings_v4.py:
```diff
@@ -0,0 +1,381 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+
+import gc
```
Can you test the correctness of the model against the HF implementation?

DarkLight1337 commented on vllm/config.py:
```diff
@@ -3258,7 +3258,8 @@ def get_limit_per_prompt(self, modality: str) -> int:
     """Controls the behavior of output pooling in pooling models."""
 
-    pooling_type: Optional[str] = None
+    pooling_type: Optional[Literal["last", "all", "cls", "step", "mean",
```
Actually the pooling type here is supposed to be upper case

DarkLight1337 commented on vllm/model_executor/layers/pooler.py:
```diff
@@ -706,3 +739,62 @@ def forward(
+    def from_config(cls, model_config: ModelConfig) -> "VisionPooler":
+        return cls(model_config)
+
+    def __init__(self, config: ModelConfig):
```
Can we pass in the token IDs and hidden size explicitly? In case other models store those attributes in different locations

DarkLight1337 commented on vllm/model_executor/models/jina_embeddings_v4.py:
```diff
@@ -0,0 +1,64 @@
+        super().__init__(vllm_config=vllm_config,
+                         prefix=maybe_prefix(prefix, "qwen2_vl"))
+
+        self.pooler = JinaVLPooler(vllm_config)
```
Why not directly use VisionPooler here?

rnddvsrt commented:
Any update on this PR? 👍
